### PR TITLE
Defensics CoAP: fixes for cases with long data packets

### PR DESCRIPTION
**Pull request by KKopyscinski**

### CODE REVIEW

**Comment by KKopyscinski** [comment]
@michal-narajowski Could you review this?

File: defensics/tcp_data_handler.py [hunk 1]
```diff
@@ -13,23 +14,48 @@ def __init__(self, proxy: CoapProxy, tcp_server: TCPServer):
...
+                        continue
+                    else:
+                        skip = False
+                    try:
+                        rsp = self.proxy.send(data)
+                        self.tcp_server.send(rsp)
+                    except (dbus.exceptions.DBusException, TimeoutError) as ex:
+                        print(ex.args, type(ex))
+                        # TimeoutError is generated when no response is received, but it isn't expected for every packet
+                        if type(ex) == dbus.exceptions.DBusException:
```
Comment (michal-narajowski) [inline defensics/tcp_data_handler.py:41]: You can handle exceptions separately.

File: defensics/tcp_data_handler.py [hunk 2]
```diff
@@ -13,23 +14,46 @@ def __init__(self, proxy: CoapProxy, tcp_server: TCPServer):
...
-        for data in self.tcp_server.recv():
-            if self.proxy.is_ready():
-                try:
-                    rsp = self.proxy.send(data)
-                    self.tcp_server.send(rsp)
-                except TimeoutError:
-                    logging.debug("Response timeout!")
-                    # because proxy isn't run in separate process, it can easily be restarted
+            for data in self.tcp_server.recv():
+                if self.proxy.is_ready():
```
Comment (michal-narajowski) [inline defensics/tcp_data_handler.py:28]: Exit early to prevent huge indented chunks of code.

File: defensics/tcp_data_handler.py [hunk 3]
```diff
@@ -13,23 +14,46 @@ def __init__(self, proxy: CoapProxy, tcp_server: TCPServer):
         logging.debug('local proxy: ' + str(self.proxy) + 'global proxy: ' + str(proxy))
 
     def run(self):
-        rc = 1
-        while rc != 0:
-            rc = self.proxy.run()
-            logging.debug('Proxy retry')
-        logging.debug(str(self.proxy.device_iface) + str(self.proxy.req_char_iface) + str(self.proxy.rsp_char_iface))
         self.tcp_server.start()
+        skip = False
```
Comment (michal-narajowski) [inline defensics/tcp_data_handler.py:18]: Overall this whole procedure is way too complicated. Refactor it into smaller methods with descriptive names. You can create a simple state machine to clearly define states. Proper error handling, less nested chunks of code.

**Review by michal-narajowski — CHANGES_REQUESTED** [review]

**Comment by KKopyscinski** [comment]
@michal-narajowski Done, I think readability is much better now

File: defensics/tcp_data_handler.py [hunk 4]
```diff
@@ -3,33 +3,75 @@
...
+    def run(self):
         self.tcp_server.start()
+        logging.debug('running data handler')
+        try:
+            self._run_proxy()
+        except dbus.exceptions.DBusException as ex:
+            self._handle_dbus_exception(ex)
 
         for data in self.tcp_server.recv():
             if self.proxy.is_ready():
```
Comment (michal-narajowski) [inline defensics/tcp_data_handler.py:60]: Exit early

File: defensics/tcp_data_handler.py [hunk 5]
```diff
@@ -3,33 +3,75 @@
...
-                    rsp = self.proxy.send(data)
-                    self.tcp_server.send(rsp)
-                except TimeoutError:
-                    logging.debug("Response timeout!")
-                    # because proxy isn't run in separate process, it can easily be restarted
-                    self.proxy.run()
-                    logging.debug("Reconnecting")
+                # skipping long data packets, as they're continuation of previously sent long data. Shorter ones
+                # might be beginning of new data, so don't ignore them and cancel skip.
+                if self._send_or_pass(data):
```
Comment (michal-narajowski) [inline defensics/tcp_data_handler.py:63]: Exit early, avoid nested code.

**Review by michal-narajowski — CHANGES_REQUESTED** [review]
I would simplify it a bit more. Having the CoaP Proxy and TCP server decoupled from the automation handler is nice but I think it may be a good idea to connect them in some way. 
How I see it:
- Each test case is separate
- In the before step, we initialize the TCP server and CoAP Proxy
- TCP server is rather stable so it should be fine but if there is an error -> fail test and tear down the entire setup
- If CoAP fails - handle exceptions properly. If an exception is unexpected -> fail the test and tear down the setup
- If it is expected (no response or disconnect) -> verify that there was no crash
- If there were no crash -> PASS -> tear down the setup
- If there was a crash -> FAIL -> capture whatever we can (optional) -> tear down the setup
- Next test case and do everything from the beginning
- In the teardown, you can flush all the additional data from the TCP server that the proxy cannot process.

I believe in this way we can avoid nested loops, exception handlers all over the place, and magic values.

**Comment by KKopyscinski** [comment]
@michal-narajowski Ok, I think I'll add another commit joining coap_automation_handler with rest of the code. As you said, running it independently is nice, but in to many places we need to interact with it. Also, ability access to tcp_server and proxy from this module simplifies a lot of things

**Comment by KKopyscinski** [comment]
This is solved by https://github.com/JuulLabs-OSS/BTPTesterCore/pull/42, also, there is no point in testing packets longer than 512, as CoAP over BLE uses GATT Writes and more cannot be sent over radio

**Pull request closed without merging**